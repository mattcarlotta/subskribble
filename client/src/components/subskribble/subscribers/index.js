import React from 'react';
import SubscriptionsPanels from './panels/subscriptionsPanels';
import PanelsContainer from '../app/panels/panelsContainer';

export default function() { return <PanelsContainer Panels={SubscriptionsPanels} />  }
