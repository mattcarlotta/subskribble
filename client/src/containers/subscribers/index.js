import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../../actions/subscriberActions';
import CARDS from '../../components/subscribers/layouts/PanelCards/panelCards';
import PanelLoader from '../../components/app/panels/PanelLoader/PanelLoader';
import SubsPanel from '../../components/subscribers/panels/subscriptionsPanels';

const Subscribers = props => (
  <PanelLoader
    {...props}
    buttonIcon="person_add"
    buttonPushLocation="subscribers/register"
    cardTitle="Subscribers"
    CARDS={CARDS}
    Panel={SubsPanel}
  />
);

export default connect(
  state => ({ serverMessage: state.server.message, ...state.subs }),
  { ...actions },
)(Subscribers);
