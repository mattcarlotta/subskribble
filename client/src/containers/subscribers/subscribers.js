import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../../actions/subscriberActions';
import CARDS from '../../components/subscribers/layouts/PanelCards/panelCards';
import PanelLoader from '../../components/app/panels/PanelLoader/PanelLoader.js';

const Subscribers = props => (
  <PanelLoader
    buttonIcon="person_add"
    buttonPushLocation="subscribers/register"
    cardTitle="Subscribers"
    CARDS={CARDS}
    panelType="tab"
    tipTitle="Add Subscriber"
    {...props}
  />
);

export default connect(
  state => ({
    serverError: state.server.error,
    serverMessage: state.server.message,
    ...state.subs,
  }),
  { ...actions },
)(Subscribers);
