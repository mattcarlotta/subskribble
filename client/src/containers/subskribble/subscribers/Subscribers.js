import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { fetchSubscribers, fetchSubscriberCounts } from '../../../actions/tableActions';
import CARDS from '../../../components/subskribble/subscribers/layouts/panelCards';
import SubscriptionsPanel from '../../../components/subskribble/subscribers/panels/subscriptionsPanels';
import Loader from '../../../containers/subskribble/app/loading/Loader';

class Subscribers extends PureComponent {
  componentDidMount = () => {
    const { activesubcount, inactivesubcount, fetchSubscriberCounts, fetchSubscribers } = this.props;
    (!activesubcount || !inactivesubcount) && fetchSubscriberCounts();
    fetchSubscribers();
  }

  render = () => (
    (!this.props.activesubs || !this.props.inactivesubs || !this.props.activesubcount || !this.props.inactivesubcount)
      ? <Loader buttonLabel="Add Subscriber" formNum={0} />
      : <SubscriptionsPanel CARDS={CARDS({...this.props})} />
  )
}

export default connect(state => ({
  activesubs: state.fields.activesubs,
  activesubcount: state.fields.activesubcount,
  inactivesubs: state.fields.inactivesubs,
  inactivesubcount: state.fields.inactivesubcount
}), { fetchSubscribers, fetchSubscriberCounts })(Subscribers)
