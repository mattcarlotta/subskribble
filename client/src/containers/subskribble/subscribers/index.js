import React from 'react';
import { connect } from 'react-redux';
import actions from '../../../actions/subscriberActions';
import CARDS from '../../../components/subskribble/subscribers/layouts/panelCards';
import PanelLoader from '../../../components/subskribble/app/panels/PanelLoader';
import SubsPanel from '../../../components/subskribble/subscribers/panels/subscriptionsPanels';

const Subscribers = props => (
  <PanelLoader
    {...props}
    buttonLabel="Add Subscriber"
    buttonPushLocation="subscribers/register"
    cardTitle="Subscribers"
    CARDS={CARDS}
    Panel={SubsPanel}
  />
);

export default connect(state => ({ ...state.subs }), { ...actions })(Subscribers)
