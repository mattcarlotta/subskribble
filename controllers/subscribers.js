const statusType = status => (status.length > 1 ? `WHERE status='${status[0]}' OR status='${status[1]}'` : `WHERE status='${status[0]}'`);

const query = {
  deleteSub: () => ("DELETE FROM subscribers WHERE id=$1 RETURNING *"),
  getSubs: (limit, offset, status) => (`SELECT id, key, status, email, subscriber, plan, startdate, enddate, amount FROM subscribers ${statusType(status)} ORDER BY key ASC LIMIT ${limit} OFFSET ${offset};`),
  getSubcount: () => (
    `SELECT count(*) filter (where status = 'active') AS active, count(*) filter (where status in ('inactive', 'suspended')) as inactive FROM subscribers;`
  ),
  updateSub: () => ("UPDATE subscribers SET status=$1, enddate=$2 WHERE id=$3 RETURNING subscriber")
}

module.exports = app => {
  const { db } = app.database;
  const { parseStringToNum } = app.shared.helpers;
  const moment = app.get("moment");

  const controller = {
    // subscribers methods
    index: (req, res) => _index(req,res),
    fetchRecords: (req, res) => _fetchRecords(req, res),
    fetchCounts: (req, res) => _fetchCounts(req, res),
    // create: (req, res) => _create(req,res)
    // show: (req, res) => _show(req,res)
    update: (req, res) => _update(req, res),
    delete: (req, res) => _delete(req, res)
  }

  const _index = async (req, res) => {
    try {
      const activesubscribers = await db.any(query.getSubs(10, 0, ['active']));
      const inactivesubscribers = await db.any(query.getSubs(10, 0, ['inactive', 'suspended']));

      res.status(201).json({ activesubscribers, inactivesubscribers });
    } catch (err) {
      return res.status(500).json({ err: err.toString() })
    }
  }

  const _delete = async (req, res) => {
    const { id } = req.params;
    try {
      const name = await db.result(query.deleteSub(), id)

      res.status(201).json({ message: `Succesfully deleted ${name.rows[0].subscriber}.` });
    } catch (err) {
      return res.status(500).json({ err: err.toString() })
    }
  }

  const _fetchRecords = async (req, res) => {
    let { table, limit, page } = req.query;
    limit = parseStringToNum(limit);
    const offset =  parseStringToNum(page) * limit;
    const status = table === 'activesubscribers' ? ['active'] : ['inactive', 'suspended'];

    try {
      let activesubscribers, inactivesubscribers;
      const subscribers = await db.any(query.getSubs(limit, offset, status));

      (table === "activesubscribers") ? activesubscribers = subscribers : inactivesubscribers = subscribers;

      res.status(201).json({ activesubscribers, inactivesubscribers });
    } catch (err) {
      return res.status(500).json({ err: err.toString() })
    }
  }

  const _fetchCounts = async (req, res) => {
    try {
      const subscribers = await db.any(query.getSubcount());

      res.status(201).json({
        activesubscriberscount: parseStringToNum(subscribers[0].active),
        inactivesubscriberscount: parseStringToNum(subscribers[0].inactive)
      });
    } catch (err) {
      return res.status(500).json({ err: err.toString() })
    }
  }

  const _update = async (req, res) => {
    const { id } = req.params;
    const { updateType, statusType } = req.body;
    const endDate = updateType === 'suspended' ? moment().format("MMM DD, YYYY") : null;

    try {
      const name = await db.one(query.updateSub(), [statusType, endDate, id])

      res.status(201).json({ message: `Succesfully ${updateType} ${name.subscriber}.` });
    } catch (err) {
      return res.status(500).json({ err: err.toString() })
    }
  }

  return controller;
}
