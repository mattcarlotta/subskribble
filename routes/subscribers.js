module.exports = app => {
  const { subscribers: { create, index, deleteOne, fetchCounts, fetchRecords, updateOne } } = app.controllers;
  const { requireAuth } = app.services.strategies;

  app.post('/api/subscribers/signup', requireAuth, create)
  app.get('/api/subscribers', requireAuth, index)
  app.get('/api/subscribercounts', requireAuth, fetchCounts)
  app.get('/api/subscribers/records', requireAuth, fetchRecords)
  // app.post('/v1/customers', auth.checkSudo, auth.isActiveUser, controller.create)
  // app.get('/v1/customers/:id', auth.checkSudo, auth.isActiveUser, controller.show)
  app.put('/api/subscribers/update/:id', requireAuth, updateOne)
  app.delete('/api/subscribers/delete?', requireAuth, deleteOne)
}
