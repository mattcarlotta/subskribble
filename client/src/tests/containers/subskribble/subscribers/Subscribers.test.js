import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Subscribers from '../../../../containers/subskribble/subscribers/Subscribers';
import { ACTIVESUBSCRIBERS, INACTIVESUBSCRIBERS } from '../../../fixtures/subscribers';

let wrapper;

describe('[SUBSCRIBERS]', () => {
  beforeEach(() => {
    wrapper = shallow(<Subscribers />);
  })

  it('should load Subscribers component', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  })

  it('should update Subscriber\'s local state and render the TableList', async () => {
    await wrapper.instance().fetchAllSubscribers();
    expect(wrapper.props().activesubscribers).toEqual(ACTIVESUBSCRIBERS);
    expect(wrapper.props().inactivesubscribers).toEqual(INACTIVESUBSCRIBERS);
    wrapper.update();
    expect(toJSON(wrapper)).toMatchSnapshot();
  })

})
