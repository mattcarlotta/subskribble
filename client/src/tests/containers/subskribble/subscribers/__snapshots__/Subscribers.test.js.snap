// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SUBSCRIBERS] should load Subscribers component 1`] = `<Component />`;

exports[`[SUBSCRIBERS] should update Subscriber's local state and render the TableList 1`] = `
<Component
  CARDS={
    Array [
      Object {
        "FILTERFIELDLABEL": "Filter Active Subscriptions",
        "FILTERFORM": "FilterActiveSubscriptions",
        "SELECTFIELDITEMS": Array [
          "10",
          "20",
          "50",
          "All",
        ],
        "TAB": "Active Subscribers",
        "TABLECONTENTS": Array [
          Object {
            "amount": "29.99",
            "email": "lukeskywalker@rebelforce.com",
            "enddate": null,
            "id": "280e5e69-23cc-11e8-9ee2-87c4367aefbc",
            "key": 6,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Siemen Walker",
          },
          Object {
            "amount": "29.99",
            "email": "bob-eh@sap.com",
            "enddate": null,
            "id": "280e5e66-23cc-11e8-9ee2-df58ddf546ce",
            "key": 3,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Bob Aronssen",
          },
          Object {
            "amount": "29.99",
            "email": "shani.smith@hotmail.com",
            "enddate": null,
            "id": "280e5e67-23cc-11e8-9ee2-f3fe7042465f",
            "key": 4,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Shaniqua Smith",
          },
          Object {
            "amount": "29.99",
            "email": "jLamar@gmail.com",
            "enddate": null,
            "id": "280e5e6a-23cc-11e8-9ee2-d78220c1be75",
            "key": 7,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Jerry Lamar",
          },
          Object {
            "amount": "29.99",
            "email": "tanyaballschin@gmail.com",
            "enddate": null,
            "id": "280e5e68-23cc-11e8-9ee2-8bf0d337d932",
            "key": 5,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Tanya Ballschin",
          },
          Object {
            "amount": "29.99",
            "email": "admin@admin.com",
            "enddate": null,
            "id": "280e5e64-23cc-11e8-9ee2-039e1dd23d0e",
            "key": 1,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Admin",
          },
          Object {
            "amount": "29.99",
            "email": "squatters@gmail.com",
            "enddate": null,
            "id": "280e5e65-23cc-11e8-9ee2-874a79b185c0",
            "key": 2,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "active",
            "subscriber": "Sherry Waters",
          },
        ],
        "TABLEHEADERS": Array [
          Object {
            "dataIndex": "status",
            "render": [Function],
            "title": "Status",
          },
          Object {
            "dataIndex": "subscriber",
            "title": "Subscriber",
          },
          Object {
            "dataIndex": "plan",
            "title": "Plan",
          },
          Object {
            "dataIndex": "startdate",
            "title": "Start Date",
          },
          Object {
            "dataIndex": "enddate",
            "title": "End Date",
          },
          Object {
            "dataIndex": "amount",
            "title": "Amount",
          },
        ],
      },
      Object {
        "FILTERFIELDLABEL": "Filter Inactive Subscriptions",
        "FILTERFORM": "FilterInactiveSubscriptions",
        "SELECTFIELDITEMS": Array [
          "10",
          "20",
          "50",
          "All",
        ],
        "TAB": "Inactive Subscribers",
        "TABLECONTENTS": Array [
          Object {
            "amount": "29.99",
            "email": "carlsagan42@yahoo.com",
            "enddate": "Jan 3, 2018",
            "id": "280e5e6b-23cc-11e8-9ee2-4bb5be2cf2fd",
            "key": 8,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "inactive",
            "subscriber": "Carl Sagan",
          },
          Object {
            "amount": "29.99",
            "email": "seamark@outlook.com",
            "enddate": "Jan 12, 2018",
            "id": "280e5e6c-23cc-11e8-9ee2-67c5e7700f99",
            "key": 9,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "inactive",
            "subscriber": "Mark Canelo",
          },
          Object {
            "amount": "29.99",
            "email": "axxll@manjaro.com",
            "enddate": "Jan 16, 2018",
            "id": "280e5e6d-23cc-11e8-9ee2-572a31e0fc7f",
            "key": 10,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "suspended",
            "subscriber": "Axle Root",
          },
          Object {
            "amount": "29.99",
            "email": "vicksAdam@sap.com",
            "enddate": "Jan 17, 2018",
            "id": "280e5e6e-23cc-11e8-9ee2-53bd93663073",
            "key": 11,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "inactive",
            "subscriber": "Adamn Vicks",
          },
          Object {
            "amount": "29.99",
            "email": "wallyworld@manjaro.com",
            "enddate": "Jan 17, 2018",
            "id": "280e5e6f-23cc-11e8-9ee2-fb171b00c1fa",
            "key": 12,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "suspended",
            "subscriber": "Wes Walls",
          },
          Object {
            "amount": "29.99",
            "email": "aleashtrails@gmail.com",
            "enddate": "Jan 17, 2018",
            "id": "280e5e70-23cc-11e8-9ee2-13fe281b583d",
            "key": 13,
            "plan": "Carlotta Prime",
            "startdate": "Mar 09, 2018",
            "status": "suspended",
            "subscriber": "Alisha Tallis",
          },
        ],
        "TABLEHEADERS": Array [
          Object {
            "dataIndex": "status",
            "render": [Function],
            "title": "Status",
          },
          Object {
            "dataIndex": "subscriber",
            "title": "Subscriber",
          },
          Object {
            "dataIndex": "plan",
            "title": "Plan",
          },
          Object {
            "dataIndex": "startdate",
            "title": "Start Date",
          },
          Object {
            "dataIndex": "enddate",
            "title": "End Date",
          },
          Object {
            "dataIndex": "amount",
            "title": "Amount",
          },
        ],
      },
    ]
  }
/>
`;
